### PR TITLE
build: transition to C++17, minor changes in documentation

### PR DESCRIPTION
**Pull request by mshabunin**
* Raise C++ standard to C++17
* Show warnings for possible configuration issues (C++ standard, GCC/Clang compiler versions), but do not halt the configuration process
  * Selected VS 2017 15.7 (1914) as a first supported version, because it should support all C++17 language features (see links)
* Raise minimum cmake version to 3.13
  * another variant is 3.10 which was bundled with VS2017, but IMO it would be better to choose higher version because it would reduce supported versions range and might have some useful functionality (e.g. `list(JOIN)` or `list(TRANSFORM)`)
* Clean-up current 2.4->3.x transition guide - it should contain information for 4.x->5.x migration in future
* Minor changes in Linux installation documentation snippets-scripts
* Disabled strange VS warnings, because we do not include any of listed headers in our code. Perhaps this is from `<complex.h>` which is used in some places, e.g. #14557
   ```
   Warning	C4996	'_Header_ccomplex': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. 
    ```

See also:
- https://learn.microsoft.com/en-us/cpp/overview/compiler-versions?view=msvc-170
- https://en.cppreference.com/w/cpp/compiler_support/17

```
force_builders=Win64 OpenCL,Custom Win
```

### CODE REVIEW

**Comment by asmorkalov** [comment]
ADE: https://github.com/opencv/opencv/pull/26353

**Comment by asmorkalov** [comment]
iOS:
```
-- FP16 is not supported by C++ compiler
-- Performing Test HAVE_CPU_BASELINE_FLAGS
-- Performing Test HAVE_CPU_BASELINE_FLAGS - Failed
CMake Error at cmake/OpenCVCompilerOptimizations.cmake:654 (message):
  Compiler doesn't support baseline optimization flags:
Call Stack (most recent call first):
  cmake/OpenCVCompilerOptions.cmake:354 (ocv_compiler_optimization_options)
  CMakeLists.txt:630 (include)
```

**Comment by mshabunin** [comment]
> iOS:
> 
> ```
> -- FP16 is not supported by C++ compiler
> -- Performing Test HAVE_CPU_BASELINE_FLAGS
> -- Performing Test HAVE_CPU_BASELINE_FLAGS - Failed
> CMake Error at cmake/OpenCVCompilerOptimizations.cmake:654 (message):
>   Compiler doesn't support baseline optimization flags:
> Call Stack (most recent call first):
>   cmake/OpenCVCompilerOptions.cmake:354 (ocv_compiler_optimization_options)
>   CMakeLists.txt:630 (include)
> ```

Since all compiler checks fail prior to this I have suspicion that cmake adds wrong compiler option in order to enable C++17. Is there a way to get build directory to check CMakeError.log? Can we check which cmake version is installed on OSX build machines?

**Comment by mshabunin** [comment]
On the other hand is :confused: :
```
/opencv/3rdparty/protobuf/src/google/protobuf/arena.h:412:14: error: aligned allocation function 
of type 'void *(std::size_t, std::align_val_t)' is only available on iOS 11 or newer
       return new Arena(internal::MessageOwned{});
              ^
```

File: cmake/OpenCVMinDepVersions.cmake [hunk 1]
```diff
@@ -1,5 +1,6 @@
 if(NOT DEFINED MIN_VER_CMAKE)
-  set(MIN_VER_CMAKE 3.7)
+  set(MIN_VER_CMAKE 3.13) # Debian 10
+  # set(MIN_VER_CMAKE 3.10) # Visual Studio 2017 15.7
```
Comment (asmorkalov) [inline cmake/OpenCVMinDepVersions.cmake:3]: Let's use the lowest possible version.

**Review by asmorkalov — APPROVED** [review]

**Comment by Kumataro** [comment]
Hello, `MINVER_CMAKE 3.13` makes CI errors at linux-1 and linux-4, they are installed cmake 3.10.2.
Currently all pull requests for 5.x branch will be failed automatically.
Please could you confirm about this ?

https://pullrequest.opencv.org/buildbot/builders/precommit_linux64?numbuilds=25

**Comment by Kumataro** [comment]
And windows-1(CI server) installed VS14. But 5.x branch requests VS17.

https://github.com/opencv/opencv/blob/8a5ec4bf7b306aba026d88432e84cd3973d49eb4/cmake/OpenCVDetectCXXCompiler.cmake#L229-L234

( from 4.x pull request)
```
--   Platform:
--     Timestamp:                   2024-10-26T16:44:54Z
--     Host:                        Windows 10.0.19045 AMD64
--     CMake:                       3.20.21032501-MSVC_2
--     CMake generator:             Visual Studio 14 2015 Win64
--     CMake build tool:            MSBuild.exe
--     MSVC:                        1900
--     Configuration:               Debug Release
--     Algorithm Hint:              ALGO_HINT_ACCURATE
```

**Comment by opencv-alalek** [comment]
> Raise minimum cmake version to 3.13

There are missing changes in samples
Step 10: http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/109440

**Comment by opencv-alalek** [comment]
More precommit issues are on builders for PR #26283

File: cmake/OpenCVDetectCXXCompiler.cmake [hunk 2]
```diff
@@ -197,39 +197,42 @@ elseif(MINGW)
...
-    if(HAVE_STD_CXX11)
-      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
-      set(HAVE_CXX11 ON)
-    endif()
+  if("cxx_std_17" IN_LIST CMAKE_CXX_COMPILE_FEATURES)
+    set(HAVE_CXX17 ON)
   endif()
 endif()
 
 if(NOT HAVE_CXX11)
```
Comment (rursprung) [inline cmake/OpenCVDetectCXXCompiler.cmake:212]: C++11 support is implied by C++17 support - so i presume `HAVE_CXX11` could be removed completely?

**Pull request merged**